### ISSUE TITLE
Parse Genbank/EMBL feature qualifiers into dict

### ISSUE THREAD

**Issue by innovate-invent** [biopython/biopython#2131]
Feature qualifiers anticodon, rpt_unit_range, tag_peptide, and transl_except values are complex data and should be parsed.
Example value:
`(pos:678..680,aa:Leu,seq:taa)`

This should be parsed into a dict:
```json
{
    'pos': CompoundLocation(678..680),
    'aa': Seq('Leu', Bio.Alphabet.ThreeLetterProtein()),
    'seq':Seq('taa'),
}
```
Parsing the `pos` element requires the same logic as the location descriptor.
This will require factoring out:
https://github.com/biopython/biopython/blob/264c68972909bf0c74cf6736fae36a703d82fe62/Bio/GenBank/__init__.py#L1016-1184

The alternative is a hack along the lines of:
```python
from Bio.Genbank import _FeatureConsumer

consumer = _FeatureConsumer()
consumer._expected_size = len(record)
consumer._cur_feature = SeqFeature()
consumer._seq_type = "circular" #TODO
consumer.location(qualifier.pos)
qualifier.pos = consumer._cur_feature.location
```

I would appreciate any input as per the state of the linked code or any alternatives to gain this functionality.

**Comment by peterjc** [comment]
It certainly would be a nice enhancement to parse these, although I'm not convinced about the dict idea:

```
{
    'pos': CompoundLocation(678..680),
    'aa': Seq('Leu', Bio.Alphabet.ThreeLetterProtein()),
    'seq':Seq('taa'),
}
```

I would certainly consider a named tuple (not available to Biopython until relatively recently due to the older versions of Python we supported), or new class. Examples of the other feature types would probably sway me.

Minor asides: In terms of the position, that doesn't look like a ``CompoundLocation`` (used for GenBank/EMBL join locations), but a simple ``FeatureLocation``. And, also ``Seq`` objects don't (currently) support the three letter alphabets, but that's a separate issue.

Another more important point is should we [or have we already] broken the rule/expectation that all the feature ``.qualifier`` dict values are lists of strings? We don't currently even turn the translation table string into an int.

If we strongly want to maintain the string behaviour, then a new class perhaps exposing the dict-like or tuple-like behaviour, but also with ``__str__`` giving the GenBank/EMBL string might be attractive.

Implementation wise, a new function turning the string into the parsed structure would be nice (could be used in isolation of the main parser as a bonus). We could then have a new qualifier-key based dictionary of value-parsing-functions, applied to all the relevant feature qualifiers? This would be extensible.

**Comment by innovate-invent** [comment]
I agree, my vote is for a class `TranslationLocation` rather than a dict. The index interface of a named tuple doesn't make sense in this context. 

`CompoundLocation` was to emphasize that these qualifiers can have the full complexity of the location descriptor format.

I copied the relevant examples from http://www.insdc.org/files/feature_table.html
```
/anticodon=(pos:<location>,aa:<amino_acid>,seq:<text>)
/rpt_unit_range=202..245
/tag_peptide=90..122
/transl_except=(pos:X22222:15..17,aa:Ala)
```

From this rpt_unit_range and tag_peptide values should be of type `FeatureLocation`/`CompoundLocation`.

anticodon and transl_except should be of type `TranslationLocation`.
`TranslationLocation.seq == None` for transl_except.

How should we proceed to factor out the location parsing?
What should the interface be given its current dependencies on _expected_size, _cur_feature, and _seq_type?
Where should this function and `TranslationLocation` be located in the code base?

**Comment by peterjc** [comment]
Probably ``Bio/SeqFeature.py`` for the class (given all the location objects are there) would be best. For the GenBank/EMBL parsing, somewhere under ``Bio/GenBank/*.py`` (guided by the location parsing).

What are your thoughts on breaking the expectation that the feature qualifier values are lists of strings? Do you think the proposed ``TranslationLocation`` object can be somewhat string-like to help with that, or is that a bad idea?

**Comment by innovate-invent** [comment]
I think it would be entirly reasonable for `TranslationLocation` to provide a string interface. The issue is doing the same for `FeatureLocation` and `CompoundLocation`.

**Comment by peterjc** [comment]
Qualifiers vs attributes is a long standing historical quirk, but reflects the NCBI terminology for feature annotations (and that Biopython's SeqRecord/SeqFeature/etc objects were initially very much targeting GenBank format).

I have not reviewed all the other parsers in SeqIO to see if the qualifier values are lists of strings holds true or not; or if we have this assertion to the test framework. However, I would expect this to be a  widely held assumed in existing code and scripts. So, if we change the values to be something still very string-like, we should be OK, but otherwise this is unfortunately constraining.

For example, the INSDC ``<base_range>`` value type as used in e.g. ``/rpt_unit_range=202..245`` does match nicely to the ``FeatureLocation`` object with simple integer positions, but that is not string-like and indeed would break code expecting a string (and not just because of the conversion to Python zero-based counting):

```
>>> from Bio.SeqFeature import FeatureLocation
>>> loc = FeatureLocation(201, 245)
>>> str(loc)  # not '202..245'
'[201:245]'
>>> len(loc)  # not len('202..245')
44
>>> loc.split("..")  # not ['202', '245']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'FeatureLocation' object has no attribute 'split'
```

In terms of data interchange, because for most things GenBank/EMBL/INSDC is the richest standard supported in SeqIO, we make parsers for other formats transform their data to match how it would be parsed from a GenBank/EMBL/INSDC file (and do the reverse on output). The details here are one of the factors which stalled the original attempts to integrate GFF3 into SeqIO (another is feature child/parent relationships, and related to that joins vs multi-part GFF3 features - although the latter was clarified in later versions of GFF3).

As to:

> The parsers are incomplete if they do not fully parse the file format according to its spec. Retaining data in strings rather than properly converting it is just lazy parsing (and not the good kind). It would be reasonable if the file formats didn't specify the conceptual value type to keep them as strings, but many formats do specify this.

When structured fields are not documented, retaining data in strings is an extremely pragmatic way forward - people not using the field can ignore it, those that need it can if needed do ad hoc parsing themselves.

http://www.insdc.org/files/feature_table.html does define ``anticodon``, ``rpt_unit_range``, ``tag_peptide `` and ``transl_except`` (although some of these are relatively new - it is likely they all post-date the original Biopython GenBank parser design).

Do any of these specific GenBank qualifiers with structured values map to similar constructs in GFF?

(Another thought, what about the XML forms of these GenBank fields?)

Replacing or supplementing the ``SeqFeatures`` existing dictionary ``.qualifiers`` (holding lists of strings) with a new dictionary ``.attributes`` (holding arbitrary objects) might be a way forward. We would need to consider one value vs multiple values, and also no values in things like the ``/pseudo`` and ``/trans_splicing`` as well.
